src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalysisScope.cs

```diff
@@ -130,14 +156,26 @@ public bool ContainsSpan(TextSpan filterSpan)
 
         public bool ShouldInclude(Diagnostic diagnostic)
         {
-            if (FilterTreeOpt == null)
+            if (FilterFileOpt == null)
             {
                 return true;
             }
 
-            if (!diagnostic.Location.IsInSource || diagnostic.Location.SourceTree != FilterTreeOpt)
+            if (diagnostic.Location.IsInSource)
             {
-                return false;
+                if (FilterFileOpt?.SourceTree == null ||
+                    diagnostic.Location.SourceTree != FilterFileOpt.SourceTree)
```

Comment:
Could simplify this entire `if` check to `if (diagnostic.Locatino.SourceTree == FilterFileOpt.SourceTree)`. The `IsInSource` check already verified `Location.SourceTree` isn't `null`